File: lib/javascript/src/journey-summary.ts
import {
	assertChecksum,
	assertInRange,
	assertLength,
	BLOCK_SIZE,
	isZero,
	withChecksum,
} from "./bytes.ts";
import { decodeDate, encodeDate, type Date16Bit } from "./date.ts";
import type { Direction } from "./transaction.ts";
import { cardTypeByte, cardTypeFromByte, type CardTypeName } from "./type.ts";

/** Block 10 of a top up card, rewritten on every journey and untouched by top ups. */
export type JourneySummary = {
	/** The journey before the current one; absent on the first ever journey. */
	previous?: { route: number; direction: Direction };
	/** When the last paid journey happened, also after a free transfer. */
	lastPaidAt: Date16Bit & { hour: number; minute: number };
	/** Same as byte 4 of the current transaction. */
	consecutivePayments: number;
	/** Product of the card. */
	cardType: CardTypeName;
	/** Whether the current journey was a free transfer. */
	free: boolean;
	/** Route id of the current journey. */
	route: number;
	/** Direction of the current journey. */
	direction: Direction;
	/** `0x63` after a paid journey, `0x62` after a free transfer. */
	transfersLeft: number;
};

/** The constant block 10 of personal cards. */
export const PERSONAL_JOURNEY_SUMMARY: Uint8Array = Uint8Array.fromHex(
	"000000000000000A000000000000000A",
);

function direction(byte: number, name: string): Direction {
	if (byte !== 1 && byte !== 2)
		throw new Error(`${name} must be 1 or 2, got ${byte}`);
	return byte;
}

/**
 * Decodes block 10 of a top up card.
 * @param block The 16-byte block.
 * @throws On the personal card constant or any other invalid content.
 */
export function decodeJourneySummary(block: Uint8Array): JourneySummary {
	assertLength(block, BLOCK_SIZE, "journey summary block");
	assertChecksum(block);
	if (!isZero(block.subarray(11, 13)) || block[14] !== 0) {
		throw new Error("journey summary bytes 11, 12 and 14 must be zero");
	}
	assertInRange("hour", block[4]!, 0, 23);
	assertInRange("minute", block[5]!, 0, 59);
	const previousRoute = block[0]!;
	return {
		previous:
			previousRoute === 0
				? undefined
				: {
						route: previousRoute,
						direction: direction(block[1]!, "previous direction"),
					},
		lastPaidAt: {
			...decodeDate(block.subarray(2, 4)),
			hour: block[4]!,
			minute: block[5]!,
		},
		consecutivePayments: block[6]!,
		cardType: cardTypeFromByte(block[7]!),
		free: block[8] === 1,
		route: block[9]!,
		direction: direction(block[10]!, "direction"),
		transfersLeft: block[13]!,
	};
}

/** Encodes a journey summary into block 10. */
export function encodeJourneySummary(summary: JourneySummary): Uint8Array {
	assertInRange("hour", summary.lastPaidAt.hour, 0, 23);
	assertInRange("minute", summary.lastPaidAt.minute, 0, 59);
	assertInRange("consecutive payments", summary.consecutivePayments, 0, 0xff);
	assertInRange("route", summary.route, 0, 0xff);
	assertInRange("transfers left", summary.transfersLeft, 0, 0xff);
	const block = new Uint8Array(BLOCK_SIZE);
	if (summary.previous) {
		assertInRange("previous route", summary.previous.route, 1, 0xff);
		block[0] = summary.previous.route;
		block[1] = summary.previous.direction;
	}
	block.set(encodeDate(summary.lastPaidAt), 2);
	block[4] = summary.lastPaidAt.hour;
	block[5] = summary.lastPaidAt.minute;
	block[6] = summary.consecutivePayments;
	block[7] = cardTypeByte(summary.cardType);
	block[8] = summary.free ? 1 : 0;
	block[9] = summary.route;
	block[10] = summary.direction;
	block[13] = summary.transfersLeft;
	return withChecksum(block);
}

File: lib/rust/src/journey_summary.rs
use crate::bytes::{Block, check_checksum, in_range, with_checksum};
use crate::card_type::CardType;
use crate::date::Date;
use crate::direction::Direction;
use crate::error::{Error, Result};
use crate::route::Route;

/// A route and the direction travelled along it.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Leg {
	/// Route of the journey.
	pub route: Route,
	/// Direction along the route.
	pub direction: Direction,
}

/// When the last paid journey happened, to the minute.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct LastPaidAt {
	/// Bytes 2 and 3.
	pub date: Date,
	/// `0` to `23`.
	pub hour: u8,
	/// `0` to `59`.
	pub minute: u8,
}

/// Block 10 of a top up card, rewritten on every journey and untouched by top ups.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct JourneySummary {
	/// The journey before the current one; absent on the first ever journey.
	pub previous: Option<Leg>,
	/// Also after a free transfer, which then points at the paid ride it belongs to.
	pub last_paid_at: LastPaidAt,
	/// Same as byte 4 of the current transaction.
	pub consecutive_payments: u8,
	/// Product of the card.
	pub card_type: CardType,
	/// Whether the current journey was a free transfer.
	pub free: bool,
	/// Route of the current journey.
	pub route: Route,
	/// Direction of the current journey.
	pub direction: Direction,
	/// `0x63` after a paid journey, `0x62` after a free transfer.
	pub transfers_left: u8,
}

impl JourneySummary {
	/// The constant block 10 of personal cards, which [`decode`](Self::decode) rejects.
	pub const PERSONAL: Block = [
		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0a, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
		0x0a,
	];

	/// Decodes block 10 of a top up card.
	///
	/// # Errors
	/// [`Error::Checksum`](crate::Error::Checksum), [`Error::NonZero`](crate::Error::NonZero), [`Error::Range`](crate::Error::Range), [`Error::Direction`](crate::Error::Direction) or [`Error::UnknownCardTypeByte`](crate::Error::UnknownCardTypeByte).
	pub fn decode(block: &Block) -> Result<Self> {
		check_checksum(block)?;
		let [
			previous_route,
			previous_direction,
			d0,
			d1,
			hour,
			minute,
			consecutive_payments,
			card_type,
			free,
			route,
			direction,
			reserved11,
			reserved12,
			transfers_left,
			reserved14,
			_,
		] = *block;
		if reserved11 != 0 || reserved12 != 0 || reserved14 != 0 {
			return Err(Error::NonZero("journey summary bytes 11, 12 and 14"));
		}
		let hour = in_range("hour", hour, 0, 23)?;
		let minute = in_range("minute", minute, 0, 59)?;
		Ok(Self {
			previous: match previous_route {
				0 => None,
				route => Some(Leg {
					route: Route(route),
					direction: Direction::try_from(previous_direction)?,
				}),
			},
			last_paid_at: LastPaidAt {
				date: Date::decode([d0, d1])?,
				hour,
				minute,
			},
			consecutive_payments,
			card_type: CardType::from_byte(card_type)?,
			free: free == 1,
			route: Route(route),
			direction: Direction::try_from(direction)?,
			transfers_left,
		})
	}

	/// Encodes into block 10 with its checksum.
	///
	/// # Errors
	/// [`Error::Range`](crate::Error::Range) for a previous route of zero, an hour above 23 or a minute above 59, or the errors of [`Date::encode`].
	pub fn encode(self) -> Result<Block> {
		let hour = in_range("hour", self.last_paid_at.hour, 0, 23)?;
		let minute = in_range("minute", self.last_paid_at.minute, 0, 59)?;
		let (previous_route, previous_direction) = match self.previous {
			None => (0, 0),
			Some(leg) => (
				in_range("previous route", leg.route.0, 1, 0xff)?,
				u8::from(leg.direction),
			),
		};
		let [d0, d1] = self.last_paid_at.date.encode()?;
		Ok(with_checksum([
			previous_route,
			previous_direction,
			d0,
			d1,
			hour,
			minute,
			self.consecutive_payments,
			self.card_type.byte(),
			u8::from(self.free),
			self.route.0,
			u8::from(self.direction),
			0,
			0,
			self.transfers_left,
			0,
			0,
		]))
	}
}
